File: src/main/java/com/github/simiacryptus/aicoder/text/IndentedText.java
package com.github.simiacryptus.aicoder.text;

import org.jetbrains.annotations.NotNull;

import java.util.Arrays;
import java.util.stream.Collectors;

/**
 * This class provides a way to store and manipulate indented text blocks.
 * <p>
 * The text block is stored as a single string, with each line separated by a newline character.
 * The indentation is stored as a separate string, which is prepended to each line when the text block is converted to a string.
 * <p>
 * The class provides a static method to convert a string to an IndentedText object.
 * This method replaces all tab characters with two spaces, and then finds the minimum indentation of all lines.
 * This indentation is then used as the indentation for the IndentedText object.
 * <p>
 * The class also provides a method to create a new IndentedText object with a different indentation.
 */
public class IndentedText implements TextBlock {

    public String getIndent() {
        return indent;
    }

    public static class Factory implements TextBlockFactory<IndentedText> {
        @Override
        public IndentedText fromString(String text) {
            return IndentedText.fromString(text);
        }

        @Override
        public boolean looksLike(String text) {
            return true;
        }
    }

    protected String indent;
    protected String textBlock[];

    public IndentedText(String indent, String... textBlock) {
        this.indent = indent;
        this.textBlock = textBlock;
    }

    public static @NotNull IndentedText fromString(String text) {
        text = text.replace("\t", TAB_REPLACEMENT);
        String indent = StringTools.getWhitespacePrefix(text.split(DELIMITER));
        return new IndentedText(indent,
                Arrays.stream(text.split(DELIMITER))
                .map(s -> StringTools.stripPrefix(s, indent))
                .toArray(String[]::new));
    }

    @Override
    public @NotNull String toString() {
        return Arrays.stream(rawString()).collect(Collectors.joining(DELIMITER + getIndent()));
    }

    public @NotNull IndentedText withIndent(String indent) {
        return new IndentedText(indent, textBlock);
    }

    @Override
    public String[] rawString() {
        return this.textBlock;
    }
}

File: src/main/kotlin/com/simiacryptus/aicoder/util/IndentedText.kt
package com.simiacryptus.aicoder.util

import com.simiacryptus.util.StringUtil

/**
 * This class provides a way to store and manipulate indented text blocks.
 *
 * The text block is stored as a single string, with each line separated by a newline character.
 * The indentation is stored as a separate string, which is prepended to each line when the text block is converted to a string.
 *
 * The class provides a companion object method to convert a string to an IndentedText object.
 * This method replaces all tab characters with two spaces, and then finds the minimum indentation of all lines.
 * This indentation is then used as the indentation for the IndentedText object.
 *
 * The class also provides a method to create a new IndentedText object with a different indentation.
 */
open class IndentedText(var indent: CharSequence, vararg val lines: CharSequence) : TextBlock {

    override fun toString(): String {
        return rawString().joinToString(TextBlock.DELIMITER + indent)
    }

    override fun withIndent(indent: CharSequence): IndentedText {
        return IndentedText(indent, *lines)
    }

    override fun rawString(): Array<out CharSequence> {
        return lines
    }

    companion object {
        /**
         * This method is used to convert a string into an IndentedText object.
         *
         * @param text The string to be converted into an IndentedText object.
         * @return IndentedText object created from the input string.
         */
        fun fromString(text: String?): IndentedText {
            val processedText = (text ?: "").replace("\t", TextBlock.TAB_REPLACEMENT.toString())
            val lines = processedText.split(TextBlock.DELIMITER)
            val indent = StringUtil.getWhitespacePrefix(*lines.toTypedArray())
            return IndentedText(indent, *lines.map { StringUtil.stripPrefix(it, indent) }.toTypedArray())
        }
    }
}